wensdolls/Carbon-Credit-Marketplace
# Carbon Credit Trading Platform Implementation

## Project Description
This PR implements a blockchain-based Carbon Credit Trading Platform using Clarity smart contracts and a React frontend. The system enables transparent and efficient carbon credit trading, including tokenization of carbon credits, automated verification of carbon offset projects, a real-time marketplace for buying and selling credits, and integration with IoT devices for monitoring carbon reduction.

## New Files
1. Smart Contracts:
   - `contracts/carbon-credit-token.clar`
   - `contracts/carbon-offset-projects.clar`
   - `contracts/carbon-credit-marketplace.clar`

2. React Components:
   - `components/CarbonCreditTrading.tsx`

3. Services:
   - `services/iot-integration.ts`
   - `services/contract-interactions.ts`

4. Tests:
   - `tests/carbon-credit-token.test.ts`
   - `tests/carbon-offset-projects.test.ts`
   - `tests/carbon-credit-marketplace.test.ts`

## Main Features Implemented

### 1. Carbon Credit Token (ERC-20 like)
- Minting of carbon credit tokens
- Transfer of tokens between accounts
- Balance checking
- Setting and retrieving token metadata URI

### 2. Carbon Offset Projects
- Registration of new carbon offset projects
- Verification of projects by contract owner
- Issuance of carbon credits to verified projects
- Retrieval of project details

### 3. Carbon Credit Marketplace
- Creation of sell listings for carbon credits
- Cancellation of listings by sellers
- Purchasing of carbon credits from active listings
- Retrieval of listing details

### 4. React Frontend
- User interface for interacting with smart contracts
- Forms for project registration, credit issuance, and marketplace interactions
- Display of project and listing information

### 5. IoT Integration
- Processing of IoT data for carbon reduction monitoring
- Automatic credit issuance based on verified carbon reduction data

## Testing Instructions
1. Ensure you have Node.js and npm installed on your system.
2. Clone the repository and navigate to the project directory.
3. Install dependencies:
